src/classes/EP_TaskDependencyCheck_TDTM.cls

```diff
@@ -0,0 +1,161 @@
+                    && newEPTask.Dependent_Task__c != oldEPTask.Dependent_Task__c)
+                        || newEPTask.Engagement_Plan_Template__c != oldEPTask.Engagement_Plan_Template__c) {
+                        listToValidate.add(newEPTask);
+                }
+            }
+            if (listToValidate.size() > 0) {
+                validateTasks(listToValidate);
+            }
+        }
+
+        return null;
+    }
+
+    /*******************************************************************************************************
+    * @description For the Tasks that are new or where the EP Template or Dependency has changed,
+    * prevent cross dependencies and mismatched templates
+    * @param listPSCNew the list of PSCs from trigger new. 
+    * @param listPSCOld the list of PSCs from trigger old.
+    * @return void.  
+    ********************************************************************************************************/
+    public static void validateTasks(List<Engagement_Plan_Task__c> listToValidate) {
+        // A set of the Engagement Plan Template Ids related to the tasks.
+        Set<Id> templateIds = new Set<Id>();        
+        // Engagement Plan Task mapped to the template and all the template's related tasks
+        Map<Id, Engagement_Plan_Template__c> mapTaskIdToTemplate = new Map<Id, Engagement_Plan_Template__c>();
+        // Engagement Plan Task mapped to tasks that are dependent on it to check for template mismatch
+        Map<Id, List<Engagement_Plan_Task__c>> mapTaskIdToChildren = new Map<Id, List<Engagement_Plan_Task__c>>();
+
+        for (Engagement_Plan_Task__c newEPTask : listToValidate) {
+            templateIds.add(newEPTask.Engagement_Plan_Template__c);
+        }
+
+        for (Engagement_Plan_Template__c template : [SELECT Id, (SELECT Id, Dependent_Task__c,
+                Dependent_Task__r.Engagement_Plan_Template__c, Engagement_Plan_Template__c FROM Engagement_Plan_Tasks__r)
+                FROM Engagement_Plan_Template__c WHERE Id IN :templateIds]) {
+            for (Engagement_Plan_Task__c templateTask : template.Engagement_Plan_Tasks__r) {
+                mapTaskIdToTemplate.put(templateTask.Id, template);
+            }
+        }
+
+        for (Engagement_Plan_Task__c childTask : [SELECT Id, Dependent_Task__c, Engagement_Plan_Template__c
+                FROM Engagement_Plan_Task__c WHERE Dependent_Task__c IN : listToValidate]) {
+            if (mapTaskIdToChildren.containsKey(childTask.Dependent_Task__c)) {
+                mapTaskIdToChildren.get(childTask.Dependent_Task__c).add(childTask);
+            } else {
+                List<Engagement_Plan_Task__c> childTasks = new List<Engagement_Plan_Task__c>();
+                childTasks.add(childTask);
+                mapTaskIdToChildren.put(childTask.Dependent_Task__c, childTasks);
+            }
+        }
+
+        for (Engagement_Plan_Task__c newEPTask : listToValidate) {
```

Comment:
I'm having a bit of trouble following the dense logic here. Some inline comments would be great, and I think simplifying the map structure as suggested above might clear this up a bit. Let me know if you want to have a call to discuss. Thanks!